sql: panic: invalid datum type given: tuple{string}, expected unknown

```
CREATE TABLE IF NOT EXISTS tab_orig AS
	SELECT
		'0.0.0.0'::INET + g AS _inet
	FROM
		generate_series(0, 0) AS g;

  WITH with_604 AS (
                    SELECT *
                      FROM (VALUES ('')) AS tab_2870 (
                                                        col_7201
                                                     )
                )
SELECT NULL
  FROM tab_orig AS tab_2874
 WHERE NULL::INET
       NOT IN (
              WITH with_606 AS (
                                SELECT *
                                  FROM (
                                        VALUES (
                                                (
                                                    SELECT
                                                        NULL
                                                )
                                               )
                                       ) AS tab_2878 (
                                                        col_7216
                                                     )
                            )
            SELECT tab_2874._inet
        );
```

```
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.DatumToEncDatum(0xc0015dfc00, 0x4d87740, 0xc00140c3c0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/encoded_datum.go:167 +0x1e2
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Next(0xc001801900, 0xc001961050, 0x3, 0x3, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:179 +0x517
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.Run(0x4d020a0, 0xc0016e3180, 0x4d2c5e0, 0xc001801900, 0x4cdc1e0, 0xc0019fe700)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go:171 +0x35
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcessorBase).Run(0xc001801900, 0x4d020a0, 0xc0016e3180)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:793 +0x92
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).startInternal.func1(0xc00111c000, 0xc0004cff50, 0x0)
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:523 +0x65
created by github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).startInternal
	/home/mjibson/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:522 +0x35b
```

This appears to be yet another apply join related panic:
```
panic: invalid datum type given: tuple{string}, expected unknown [recovered]
	panic: panic while executing 1 statements: WITH _ AS (SELECT * FROM (VALUES (_)) AS _ (_)) SELECT _ FROM _ AS _ WHERE _::INET NOT IN (WITH _ AS (SELECT * FROM (VALUES ((SELECT _))) AS _ (_)) SELECT _._); caused by invalid datum type given: tuple{string}, expected unknown

goroutine 1111 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc00183f000, 0x856afc0, 0xc001068980, 0x75a0b20, 0xc0011274d0)
	/Users/yahoryuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:730 +0x330
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc00183f000, 0x856afc0, 0xc001068980)
	/Users/yahoryuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:442 +0x61
panic(0x75a0b20, 0xc0011274d0)
	/usr/local/opt/go/libexec/src/runtime/panic.go:522 +0x1b5
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.DatumToEncDatum(0xc00131fd00, 0x85f1240, 0xc0007226f0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/yahoryuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/encoded_datum.go:167 +0x1e2
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Next(0xc001614a00, 0xc00183f358, 0x856c100, 0xc0013de300, 0x0)
	/Users/yahoryuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:179 +0x517
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.Run(0x856afc0, 0xc001972ac0, 0x8596000, 0xc001614a00, 0x8545100, 0xc0006cb880)
	/Users/yahoryuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go:171 +0x35
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcessorBase).Run(0xc001614a00, 0x856afc0, 0xc001972ac0)
	/Users/yahoryuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:793 +0x92
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Run(0xc001993320, 0x856b080, 0xc0013de570, 0x7d4c0b8, 0x0, 0x0)
	/Users/yahoryuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:584 +0x1ec
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc000130780, 0xc0018a1f20, 0xc00140f290, 0xc00234bb58, 0xc001373b80, 0xc001642480, 0x0, 0x0)
	/Users/yahoryuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:337 +0x381
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).planAndRunSubquery(0xc000130780, 0x856afc0, 0xc0019725c0, 0x1, 0xc0013de330, 0x3, 0x1, 0x856c100, 0xc0013de300, 0x0, ...)
	/Users/yahoryuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:880 +0x7c8
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunSubqueries(0xc000130780, 0x856afc0, 0xc0019725c0, 0xc00183f358, 0xc00234be18, 0xc0004e1180, 0x2, 0x2, 0xc001373600, 0xc00183f101, ...)
	/Users/yahoryuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:784 +0x1ba
github.com/cockroachdb/cockroach/pkg/sql.(*applyJoinNode).runRightSidePlan(0xc0014e0600, 0x856afc0, 0xc0019725c0, 0xc0015c2d80, 0xc00183f358, 0xc0013da0c0, 0x0, 0x0)
	/Users/yahoryuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/apply_join.go:310 +0x26b
github.com/cockroachdb/cockroach/pkg/sql.(*applyJoinNode).Next(0xc0014e0600, 0x856afc0, 0xc0019725c0, 0xc0015c2d80, 0xc00183f358, 0x0, 0x856afc0, 0xc0019725c0)
	/Users/yahoryuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/apply_join.go:280 +0x6df
github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Next(0xc001614000, 0xc00183f358, 0x856c0c0, 0xc0014e0600, 0x0)
...
```
I reduced the reproduction (updated the first comment above).
cc @solongordon

--- a/pkg/sql/opt/memo/statistics_builder.go
+++ b/pkg/sql/opt/memo/statistics_builder.go
@@ -852,8 +852,10 @@ func (sb *statisticsBuilder) buildJoin(
 			s.RowCount = leftStats.RowCount
 
 		case opt.AntiJoinOp, opt.AntiJoinApplyOp:
-			s.RowCount = 0
-			s.Selectivity = 0
+			// Don't set the row count to 0 since we can't guarantee that the
+			// cardinality is 0.
+			s.RowCount = epsilon
+			s.Selectivity = epsilon
 		}
 		return
 	}

--- a/pkg/sql/opt/memo/testdata/logprops/join
+++ b/pkg/sql/opt/memo/testdata/logprops/join
@@ -1417,3 +1417,16 @@ project
  │    └── filters (true)
  └── projections
       └── variable: mn.m [type=int, outer=(13)]
+
+# Regression test #40456.
+opt
+SELECT NULL
+FROM uv
+WHERE NOT EXISTS(SELECT uv.u);
+----
+values
+ ├── columns: "?column?":5(unknown!null)
+ ├── cardinality: [0 - 0]
+ ├── key: ()
+ ├── fd: ()-->(5)
+ └── prune: (5)

--- a/pkg/sql/opt/memo/testdata/stats/join
+++ b/pkg/sql/opt/memo/testdata/stats/join
@@ -1380,3 +1380,65 @@ full-join (hash)
            │                   ├── fd: ()-->(7)
            │                   └── (NULL,) [type=tuple{unknown}]
            └── false [type=bool]
+
+expr
+(SemiJoin
+    (Values
+      [ (Tuple [ (Const 1) (Const 2) ] "tuple{int}" ) ]
+      [ (Cols [ (NewColumn "a" "int") (NewColumn "b" "int") ]) ]
+    )
+    (Scan [ (Table "uv") (Cols "u,v,rowid") ])
+    []
+    []
+)
+----
+semi-join (hash)
+ ├── columns: a:1(int!null) b:2(int!null)
+ ├── cardinality: [0 - 1]
+ ├── stats: [rows=1]
+ ├── key: ()
+ ├── fd: ()-->(1,2)
+ ├── values
+ │    ├── columns: a:1(int!null) b:2(int!null)
+ │    ├── cardinality: [1 - 1]
+ │    ├── stats: [rows=1]
+ │    ├── key: ()
+ │    ├── fd: ()-->(1,2)
+ │    └── (1, 2) [type=tuple{int}]
+ ├── scan uv
+ │    ├── columns: u:3(int) v:4(int!null) rowid:5(int!null)
+ │    ├── stats: [rows=10000]
+ │    ├── key: (5)
+ │    └── fd: (5)-->(3,4)
+ └── filters (true)
+
+expr
+(AntiJoin
+    (Values
+      [ (Tuple [ (Const 1) (Const 2) ] "tuple{int}" ) ]
+      [ (Cols [ (NewColumn "a" "int") (NewColumn "b" "int") ]) ]
+    )
+    (Scan [ (Table "uv") (Cols "u,v,rowid") ])
+    []
+    []
+)
+----
+anti-join (hash)
+ ├── columns: a:1(int!null) b:2(int!null)
+ ├── cardinality: [0 - 1]
+ ├── stats: [rows=1e-10]
+ ├── key: ()
+ ├── fd: ()-->(1,2)
+ ├── values
+ │    ├── columns: a:1(int!null) b:2(int!null)
+ │    ├── cardinality: [1 - 1]
+ │    ├── stats: [rows=1]
+ │    ├── key: ()
+ │    ├── fd: ()-->(1,2)
+ │    └── (1, 2) [type=tuple{int}]
+ ├── scan uv
+ │    ├── columns: u:3(int) v:4(int!null) rowid:5(int!null)
+ │    ├── stats: [rows=10000]
+ │    ├── key: (5)
+ │    └── fd: (5)-->(3,4)
+ └── filters (true)

--- a/pkg/sql/opt/norm/rules/join.opt
+++ b/pkg/sql/opt/norm/rules/join.opt
@@ -325,6 +325,16 @@
 =>
 $left
 
+# SimplifyZeroCardinalitySemiJoin converts a SemiJoin operator to an empty
+# Values when it's known that the right input never returns any rows.
+[SimplifyZeroCardinalitySemiJoin, Normalize]
+(SemiJoin | SemiJoinApply
+    $left:*
+    $right:* & (HasZeroRows $right)
+)
+=>
+(ConstructEmptyValues (OutputCols $left))
+
 # EliminateAntiJoin discards an AntiJoin operator when it's known that the right
 # input never returns any rows.
 [EliminateAntiJoin, Normalize]
@@ -334,6 +344,18 @@ $left
 )
 =>
 $left
+
+# SimplifyZeroCardinalityAntiJoin converts an AntiJoin operator to an empty
+# Values when it's known that the right input never returns zero rows, and
+# there is no join condition.
+[SimplifyZeroCardinalityAntiJoin, Normalize]
+(AntiJoin | AntiJoinApply
+    $left:*
+    $right:* & ^(CanHaveZeroRows $right)
+    []
+)
+=>
+(ConstructEmptyValues (OutputCols $left))
 
 # EliminateJoinNoColsLeft eliminates an InnerJoin with a one row, zero column
 # left input set. These can be produced when a Values, scalar GroupBy, or other

--- a/pkg/sql/opt/norm/testdata/rules/join
+++ b/pkg/sql/opt/norm/testdata/rules/join
@@ -1841,6 +1841,27 @@ scan a
  ├── key: (1)
  └── fd: (1)-->(2-5)
 
+opt expect=EliminateSemiJoin
+SELECT * FROM a WHERE EXISTS(VALUES (k))
+----
+scan a
+ ├── columns: k:1(int!null) i:2(int) f:3(float!null) s:4(string) j:5(jsonb)
+ ├── key: (1)
+ └── fd: (1)-->(2-5)
+
+# --------------------------------------------------
+# SimplifyZeroCardinalitySemiJoin
+# --------------------------------------------------
+# TODO(justin): figure out if there's a good way to make this still apply.
+opt disable=SimplifyZeroCardinalityGroup expect=SimplifyZeroCardinalitySemiJoin
+SELECT * FROM a WHERE EXISTS(SELECT * FROM (VALUES (k)) OFFSET 1)
+----
+values
+ ├── columns: k:1(int!null) i:2(int!null) f:3(float!null) s:4(string!null) j:5(jsonb!null)
+ ├── cardinality: [0 - 0]
+ ├── key: ()
+ └── fd: ()-->(1-5)
+
 # --------------------------------------------------
 # EliminateAntiJoin
 # --------------------------------------------------
@@ -1852,6 +1873,27 @@ scan a
  ├── columns: k:1(int!null) i:2(int) f:3(float!null) s:4(string) j:5(jsonb)
  ├── key: (1)
  └── fd: (1)-->(2-5)
+
+# --------------------------------------------------
+# SimplifyZeroCardinalityAntiJoin
+# --------------------------------------------------
+opt expect=SimplifyZeroCardinalityAntiJoin
+SELECT * FROM a WHERE NOT EXISTS(SELECT count(*) FROM b WHERE x=k)
+----
+values
+ ├── columns: k:1(int!null) i:2(int!null) f:3(float!null) s:4(string!null) j:5(jsonb!null)
+ ├── cardinality: [0 - 0]
+ ├── key: ()
+ └── fd: ()-->(1-5)
+
+opt expect=SimplifyZeroCardinalityAntiJoin
+SELECT * FROM a WHERE NOT EXISTS(VALUES (k))
+----
+values
+ ├── columns: k:1(int!null) i:2(int!null) f:3(float!null) s:4(string!null) j:5(jsonb!null)
+ ├── cardinality: [0 - 0]
+ ├── key: ()
+ └── fd: ()-->(1-5)
 
 # --------------------------------------------------
 # EliminateJoinNoColsLeft

